File: twopointers/0611-valid-triangle-number.go
package twopointers

import "sort"

// nums[i] <= nums[j] <= nums[k]
// so nums[k] + 'any one' becomes greater than 'another one'

// https://leetcode.com/problems/valid-triangle-number/

// Approach #2: Two Pointers
// Time: O(nn)
// Space: O(1)
func triangleNumber(nums []int) int {
	sort.Ints(nums)

	var total, start, end int
	for i, n := 1, len(nums); i < n; i++ {
		start, end = 0, i-1
		for start < end {
			if nums[start]+nums[end] <= nums[i] {
				start++
			} else {
				total += end - start
				end--
			}
		}
	}

	return total
}

// // Approach #1: Binary Search
// // Time: O(nnlogn)
// // Space: O(1)
// func triangleNumber(nums []int) int {
// 	sort.Ints(nums)

// 	bisectRight := func(combin int, left int, right int) int {
// 		var k int
// 		for left <= right {
// 			k = left + (right-left)/2
// 			if combin <= nums[k] {
// 				right = k - 1
// 			} else {
// 				left = k + 1
// 			}
// 		}

// 		return left
// 	}

// 	var total int

// 	for i, n := 0, len(nums); i < n; i++ {
// 		for j := i + 1; j < n; j++ {
// 			total += bisectRight(nums[i]+nums[j], j+1, n-1) - j - 1
// 		}
// 	}

// 	return total
// }

File: twopointers/0611-valid-triangle-number.py
# from bisect import bisect_left
from typing import List
import unittest

# https://leetcode.com/problems/valid-triangle-number/

# python3 -m unittest twopointers/0611-valid-triangle-number.py


class Solution(unittest.TestCase):
    # # Approach #1: Binary Search
    # # Time: O(nnlogn)
    # # Space: O(1)
    # def triangleNumber(self, nums: List[int]) -> int:
    #     nums.sort()
    #     n = len(nums)
    #     return sum(bisect_left(nums, nums[i] + nums[j], j + 1) - j - 1 for i in range(n - 2) for j in range(i + 1, n - 1))

    # Approach #2: Two Pointers
    # Time: O(nn)
    # Space: O(1)
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        n, total = len(nums), 0
        for i in range(1, n):
            start, end = 0, i - 1
            while start < end:
                if nums[start] + nums[end] > nums[i]:
                    total += end - start
                    end -= 1
                else:
                    start += 1
        return total

    def test(self):
        for nums, expected in [
            ([2, 2, 3, 4], 3),
            ([4, 2, 3, 4], 4),
        ]:
            output = self.triangleNumber(nums)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
